File: leetcode/0053/53.py
'''The thought follows a simple rule:
If the sum of a subarray is positive, it has possible to make the next value bigger, so we keep do it until it turn to negative.
If the sum is negative, it has no use to the next element, so we break.
it is a game of sum, not the elements.'''


nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]


def Solution(nums: list[int]):
    for i in range(1, len(nums)):
        if nums[i - 1] > 0:
            nums[i] += nums[i - 1]
            print(nums)
    return max(nums)


print(Solution(nums))

File: leetcode/0053/53.cpp
// Idea is very simple. Basically, keep adding each integer to the sequence until the sum drops below 0.
// If sum is negative, that should reset the sequence.
#include <iostream>
#include <vector>

class Solution
{
public:
    int maxSubArray(int A[], int n)
    {
        int ans = A[0], i, j, sum = 0;
        for (i = 0; i < n; i++)
        {
            sum += A[i];
            ans = max(sum, ans);
            sum = max(sum, 0);
        }
        return ans;
    }
};

int main(int argc, char const *argv[])
{
    Solution sol;
    return 0;
}
